fairseq/modules/moe_layer.py

```diff
@@ -0,0 +1,464 @@
+
+    def __init__(self, args, shared_moe_experts=None):
+        super().__init__()
+        self.num_workers = distributed_utils.get_data_parallel_world_size()
+        expert_centroids = torch.empty(self.num_workers, args.decoder_embed_dim)
+        torch.nn.init.orthogonal_(expert_centroids, gain=0.1)
+        self.register_parameter("expert_centroids", torch.nn.Parameter(expert_centroids))
+        self.expert_network = (
+            shared_moe_experts if shared_moe_experts
+            else nn.Sequential(*([MoESublayer(args) for _ in range(args.base_sublayers)]))
+        )
+        self.expert_id = distributed_utils.get_data_parallel_rank()
+        self.shuffle = args.base_shuffle
+        self.bloss_type = args.moe_bloss_type
+        self.bloss_weight = args.moe_bloss_weight
+
+        # Add a special attribute to the expert parameters, so we know not to sync their gradients
+        # 
+        for param in self.expert_network.parameters():
+            param.expert = True
+
+    def forward(self, input_features, *args, **kwargs):
+        features = input_features.reshape(-1, input_features.size(-1))
+        is_training = input_features.requires_grad
+
+        if self.shuffle and is_training:
+            # Send each token to a random worker, to break correlations within the batch
+            shuffle_sort = torch.randperm(features.size(0), device=features.device)
+            features = All2All.apply(features[shuffle_sort])
+
+        with torch.no_grad():
+            # Compute similarity of each token to each expert, for routing
+            token_expert_affinities = features.matmul(self.expert_centroids.transpose(0, 1))
+            # TODO @margsli should be identical to below, is the einsum faster?
+            # check that features and expert_centroids' dimensions are ordered correctly
+            # token_expert_affinities = torch.einsum('sbh,eh->sbe', features, self.expert_centroids)
+            # TODO @margsli this is only argmax, should the other sampling strategies be implemented?
+            self.last_p = torch.softmax(token_expert_affinities, -1)
+
+        # Compute which token goes to which expert
+        routing_strategy = 'greedy'
+        sort_by_expert, input_splits, output_splits = self.assignment(token_expert_affinities, strategy=routing_strategy)
+        # Swap these tokens for the right ones for our expert
+        routed_features = All2All.apply(features[sort_by_expert], output_splits, input_splits)
+
+        if routed_features.size(0) > 0:
+            # Mix in the expert network based on how appropriate it is for these tokens
+            alpha = torch.sigmoid(routed_features.mv(self.expert_centroids[self.expert_id])).unsqueeze(1)
+            routed_features = alpha * self.expert_network(routed_features) + (1 - alpha) * routed_features
+        # Return to original worker and ordering
+        result = All2All.apply(routed_features, input_splits, output_splits)[self.inverse_sort(sort_by_expert)]
```

Comment:
Can you run me through the logic behind this? I cannot quite follow if there is any difference in terms of order for (1) All2All and (2) inverse_sort.

Reply:
I think I get it now. You reverse the input and output splits and that reverses the indices and should yield the correct all2all mapping. Afterward, you reorder the sorted all2all output to its original order. Is that correct?